File: cascade/logfmt.py
"""Deterministic record grammar for the cascade's non-deterministic logs.

The human tee log (runs/cascade.log) is for eyeballs / `tail -f` and is NOT
safely parseable: a model answer that emits a line looking like a timestamp
breaks the regex scraper. This module defines a SECOND, structured stream
(runs/cascade.rec) whose grammar frames every variable-length, untrusted
payload by BYTE LENGTH, so the tokenizer never has to guess where a value
ends -- parsing is fully deterministic regardless of what the model emitted.

GRAMMAR  (the whole format -- EBNF; encoding = UTF-8)
----------------------------------------------------------------------------
    stream   = { record } ;
    record   = begin , { field } , end ;
    begin    = "%%REC" , SP , "v" , uint , SP , uint , LF ;  (* version, seq *)
    field    = key , SP , uint , LF , octet * uint , LF ;    (* uint = #bytes *)
    end      = "%%END" , LF ;
    key      = ( ALPHA | "_" ) , { ALPHA | DIGIT | "_" } ;
    uint     = DIGIT , { DIGIT } ;
    SP = %x20 ; LF = %x0A ;

Determinism rule: a `field` declares the exact byte length of its value, so
the reader consumes precisely that many bytes -- the value may contain LF,
"%%REC", "%%END", fake "12:34:56" timestamps, anything -- without ambiguity.
The only structural tokens are the line-oriented `begin`/`end`/key headers,
and a value can never be mistaken for one because its span is counted, not
delimited. Tokenizing + parsing is therefore O(n) and total (no backtracking,
no regex, no escaping).

The format is byte-framed, so this module is bytes-native: `dump_record`
returns `bytes` (encode the value once) and `parse_stream` consumes `bytes`
(no decode->re-encode round trip on the caller's file read). `parse_stream`
takes an optional `keep` projection: an unwanted field's value is skipped by
advancing the counted span WITHOUT decoding/materialising it -- but its length
header and LF terminator are still validated, so a malformed skipped field can
never desync the stream (the determinism guarantee holds under projection).

Records are append-only; a truncated trailing record (writer crashed mid-emit)
is dropped by the parser, never partially yielded.
"""
from __future__ import annotations

VERSION = 1
_BEGIN = b"%%REC"
_END = b"%%END"


def dump_record(seq: int, fields: dict[str, str]) -> bytes:
    """Serialise one record to bytes. Field order is preserved (insertion
    order); keys must match /[A-Za-z_][A-Za-z0-9_]*/ and contain no space (they
    label a length-framed value, so the value itself is unconstrained). Each
    value is UTF-8 encoded exactly once."""
    out = bytearray(b"%s v%d %d\n" % (_BEGIN, VERSION, seq))
    for key, value in fields.items():
        if not key or " " in key or "\n" in key:
            raise ValueError(f"illegal field key: {key!r}")
        body = value.encode("utf-8")
        out += b"%s %d\n" % (key.encode("utf-8"), len(body))
        out += body
        out += b"\n"
    out += _END + b"\n"
    return bytes(out)


def parse_stream_incremental(
    data: bytes, start_offset: int = 0, keep: frozenset[str] | None = None,
) -> tuple[list[dict[str, str]], int]:
    """Resumable parse. Returns `(records, next_offset)`.

    `next_offset` is the byte position from which a future call can safely
    resume: it points either past the last fully-completed record's
    `%%END\\n` / past a consumed garbage line, or BACK to the start of an
    incomplete trailing record so the next call retries it once more bytes
    arrive. This makes O(new bytes) incremental reads of an append-only
    `.rec` correct: read from `next_offset`, parse, repeat.

    Two failure modes are distinguished internally:
      * truncation (incomplete header line, or declared value length runs
        past EOF) -- the file is still being written; rewind `next_offset`
        to the record's start so it is retried later.
      * permanent corruption (no space in field header, non-numeric length,
        bad LF terminator with full bytes present) -- the outer
        garbage-line skip self-corrects (consumes line by line); `next_offset`
        advances past each consumed line.

    `keep`: when given, only fields whose key is in the set are decoded and
    stored; other fields are skipped by advancing past their counted value
    (no decode, no allocation) -- their length header and LF terminator are
    still validated, so projection cannot desync the stream. "_seq" is
    always present (it is header-derived, not a field).
    """
    records: list[dict[str, str]] = []
    i = start_offset
    safe = start_offset
    n = len(data)
    while i < n:
        record_start = i
        nl = data.find(b"\n", i)
        if nl == -1:
            break  # incomplete header line at tail; safe stays at record_start
        line = data[i:nl]
        if not line.startswith(_BEGIN + b" "):
            i = nl + 1
            safe = i  # garbage line consumed
            continue
        parts = line.split(b" ")
        # %%REC vN SEQ  -> exactly 3 tokens; malformed header => skip the line
        if len(parts) != 3 or not parts[1].startswith(b"v"):
            i = nl + 1
            safe = i  # malformed begin treated as garbage; consumed
            continue
        rec: dict[str, str] = {"_seq": parts[2].decode("utf-8", "replace")}
        i = nl + 1
        truncated = False
        ok = True
        while True:
            nl = data.find(b"\n", i)
            if nl == -1:
                truncated = True  # mid-record, no further LF yet
                break
            head = data[i:nl]
            if head == _END:
                i = nl + 1
                break
            sp = head.rfind(b" ")
            if sp == -1 or not head[sp + 1:].isdigit():
                ok = False  # permanent: bad field header
                break
            key = head[:sp].decode("utf-8", "replace")
            length = int(head[sp + 1:])
            vstart = nl + 1
            vend = vstart + length
            if vend + 1 > n:
                truncated = True  # declared length runs past current EOF
                break
            if data[vend:vend + 1] != b"\n":
                ok = False  # permanent: full bytes present, terminator wrong
                break
            if keep is None or key in keep:
                rec[key] = data[vstart:vend].decode("utf-8", "replace")
            i = vend + 1
        if truncated:
            # rewind so the next call retries this record once data grows
            i = record_start
            break
        if ok:
            records.append(rec)
            safe = i  # past `%%END\n`
        # else (permanent): the outer-while garbage-line skip self-corrects;
        # `safe` advances one line at a time as those lines are consumed.
    return records, safe


def parse_stream(
    data: bytes, keep: frozenset[str] | None = None,
) -> list[dict[str, str]]:
    """Tokenise a record stream deterministically -- thin wrapper over
    `parse_stream_incremental` for callers that do not need a resume offset.

    Unknown/garbage lines between records are skipped; an incomplete trailing
    record is dropped. See `parse_stream_incremental` for the resumable
    variant and the full semantics of `keep` (field projection)."""
    return parse_stream_incremental(data, 0, keep)[0]

File: dashboard/src/lib/logfmt.ts
/**
 * TS port of cascade/logfmt.py — bytes-native incremental parser for the
 * cascade's `.rec` record grammar.
 *
 * GRAMMAR (EBNF, UTF-8):
 *     stream = { record } ;
 *     record = begin , { field } , end ;
 *     begin  = "%%REC" , SP , "v" , uint , SP , uint , LF ;
 *     field  = key , SP , uint , LF , octet * uint , LF ;
 *     end    = "%%END" , LF ;
 *
 * Determinism rule: a field declares the exact byte length of its value, so
 * the value may contain LF / "%%REC" / "%%END" without ambiguity. Parsing is
 * O(n), total, no backtracking. A truncated trailing record is dropped (and
 * its byte offset returned as `nextOffset` so the next call retries it once
 * more bytes arrive).
 *
 * This module mirrors `cascade/logfmt.py::parse_stream_incremental` byte-for-
 * byte; see the Python source for the canonical semantics and the test suite
 * for the cross-check.
 */

const BEGIN = encodeAscii("%%REC ");
const END = encodeAscii("%%END");
const LF = 0x0a;
const SP = 0x20;
const DIGIT_0 = 0x30;
const DIGIT_9 = 0x39;
const TEXT_DECODER = new TextDecoder("utf-8", { fatal: false });

export function parseStreamIncremental(
  data: Uint8Array,
  startOffset: number = 0,
  keep?: ReadonlySet<string>,
): { records: Record<string, string>[]; nextOffset: number } {
  const records: Record<string, string>[] = [];
  let i = startOffset;
  let safe = startOffset;
  const n = data.length;

  while (i < n) {
    const recordStart = i;
    let nl = indexOfLF(data, i, n);
    if (nl === -1) {
      // incomplete header line at tail; `safe` stays at recordStart.
      break;
    }
    const line = data.subarray(i, nl);
    if (!startsWithBytes(line, BEGIN)) {
      i = nl + 1;
      safe = i; // garbage line consumed
      continue;
    }
    // Header is exactly three space-separated tokens: %%REC v<N> <seq>.
    // The split is over the already-known-short header line; allocation cost
    // is bounded by the header width, not the record value width.
    const headerParts = splitBytesBySpace(line);
    if (
      headerParts.length !== 3 ||
      headerParts[1] === undefined ||
      headerParts[1][0] !== /* 'v' */ 0x76 ||
      headerParts[2] === undefined
    ) {
      i = nl + 1;
      safe = i; // malformed begin treated as garbage
      continue;
    }
    const rec: Record<string, string> = {
      _seq: TEXT_DECODER.decode(headerParts[2]),
    };
    i = nl + 1;
    let truncated = false;
    let ok = true;
    while (true) {
      nl = indexOfLF(data, i, n);
      if (nl === -1) {
        truncated = true; // mid-record, no further LF yet
        break;
      }
      const head = data.subarray(i, nl);
      if (equalBytes(head, END)) {
        i = nl + 1;
        break;
      }
      const sp = lastIndexOfSpace(head);
      if (sp === -1 || !isDigitSpan(head, sp + 1, head.length)) {
        ok = false; // permanent: bad field header
        break;
      }
      const key = TEXT_DECODER.decode(head.subarray(0, sp));
      const length = parseDigits(head, sp + 1, head.length);
      const vstart = nl + 1;
      const vend = vstart + length;
      if (vend + 1 > n) {
        truncated = true; // declared length runs past current EOF
        break;
      }
      if (data[vend] !== LF) {
        ok = false; // permanent: full bytes present, terminator wrong
        break;
      }
      if (!keep || keep.has(key)) {
        rec[key] = TEXT_DECODER.decode(data.subarray(vstart, vend));
      }
      i = vend + 1;
    }
    if (truncated) {
      // rewind so the next call retries this record once data grows
      i = recordStart;
      break;
    }
    if (ok) {
      records.push(rec);
      safe = i; // past `%%END\n`
    }
    // else (permanent corruption): the outer-while garbage-line skip self-
    // corrects; `safe` advances one line at a time as those lines are consumed.
  }
  return { records, nextOffset: safe };
}

export function parseStream(
  data: Uint8Array,
  keep?: ReadonlySet<string>,
): Record<string, string>[] {
  return parseStreamIncremental(data, 0, keep).records;
}

/** Serialise one record back to wire bytes -- mirror of
 * `cascade.logfmt.dump_record`. The Python writer is the CANONICAL one used
 * by the cascade itself (mcp_servers/_rec.py); this TS writer exists for
 * offline tooling (seed_replay.ts, tests) where we want a writer that lives
 * in the same module as the parser it round-trips against. Field-order is
 * insertion order; keys must match the parser's accepted shape (no spaces,
 * no LF). */
export function dumpRecord(
  seq: number,
  fields: Record<string, string>,
): Uint8Array {
  const chunks: Uint8Array[] = [];
  chunks.push(ENCODER.encode(`%%REC v1 ${String(seq)}\n`));
  for (const [key, value] of Object.entries(fields)) {
    if (key === "" || /[ \n]/.test(key)) {
      throw new Error(`illegal field key: ${JSON.stringify(key)}`);
    }
    const body = ENCODER.encode(value);
    chunks.push(ENCODER.encode(`${key} ${String(body.length)}\n`));
    chunks.push(body);
    chunks.push(ENCODER.encode("\n"));
  }
  chunks.push(ENCODER.encode("%%END\n"));
  return concat(chunks);
}

const ENCODER = new TextEncoder();

function concat(chunks: Uint8Array[]): Uint8Array {
  let n = 0;
  for (const c of chunks) n += c.length;
  const out = new Uint8Array(n);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}

function encodeAscii(s: string): Uint8Array {
  const out = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
  return out;
}

function indexOfLF(data: Uint8Array, from: number, to: number): number {
  for (let i = from; i < to; i++) {
    if (data[i] === LF) return i;
  }
  return -1;
}

function startsWithBytes(haystack: Uint8Array, needle: Uint8Array): boolean {
  if (haystack.length < needle.length) return false;
  for (let i = 0; i < needle.length; i++) {
    if (haystack[i] !== needle[i]) return false;
  }
  return true;
}

function equalBytes(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function splitBytesBySpace(data: Uint8Array): Uint8Array[] {
  const parts: Uint8Array[] = [];
  let start = 0;
  for (let i = 0; i < data.length; i++) {
    if (data[i] === SP) {
      parts.push(data.subarray(start, i));
      start = i + 1;
    }
  }
  parts.push(data.subarray(start));
  return parts;
}

function lastIndexOfSpace(data: Uint8Array): number {
  for (let i = data.length - 1; i >= 0; i--) {
    if (data[i] === SP) return i;
  }
  return -1;
}

function isDigitSpan(data: Uint8Array, from: number, to: number): boolean {
  if (from >= to) return false; // empty digit span is not numeric
  for (let i = from; i < to; i++) {
    const b = data[i];
    if (b === undefined || b < DIGIT_0 || b > DIGIT_9) return false;
  }
  return true;
}

function parseDigits(data: Uint8Array, from: number, to: number): number {
  let n = 0;
  for (let i = from; i < to; i++) {
    n = n * 10 + ((data[i] ?? 0) - DIGIT_0);
  }
  return n;
}
